File: utils_local.R
suppressPackageStartupMessages({
  library(dplyr)
  library(processx)
})

.this_file_path <- tryCatch(dirname(sys.frame(1)$ofile), error = function(e) ".")
.MONITOR_MEM_SH_PATH <- file.path(.this_file_path, "monitor_mem.sh")

POLLING_INTERVAL <- '0.05'

.TIME_CONVERSIONS <- c(
  "s" = 1, "ms" = 1000, "us" = 1e6, "\u00b5s" = 1e6,
  "ns" = 1e9, "m" = 1/60, "h" = 1/3600, "d" = 1/86400
)

MemoryTimer = function(silent = TRUE, csv_path = NULL, csv_columns = NULL,
                       unit = "s", summary_unit = NULL) {
  env = environment()
  env$timings = list()
  pid = Sys.getpid()
  env$.shutdown_done = FALSE
  env$.csv_path = csv_path
  env$.csv_columns = csv_columns
  env$.unit = unit
  env$.summary_unit = if (!is.null(summary_unit)) summary_unit else unit

  with_timer = function(message, expr) {
    if (!silent) cat(paste0(message, '...\n'))
    result = NULL
    aborted = FALSE
    # Write monitor output to a temp file instead of a pipe. R is single-
    # threaded: during long-running expressions (e.g. rep() for 150+ GiB),
    # the 64 KB pipe buffer fills and the monitor blocks on write, missing
    # the peak RSS. A file has no such limit.
    tmpf <- tempfile(pattern = "memmon_", fileext = ".csv")
    curr_process <- process$new(
      command = .MONITOR_MEM_SH_PATH,
      args = c("-p", as.character(pid), "-i", POLLING_INTERVAL),
      stdout = tmpf
    )
    # Sync: wait for first sample before timing (file must have content)
    deadline <- proc.time()[['elapsed']] + 5
    repeat {
      if (file.exists(tmpf) && file.info(tmpf)$size > 0) break
      if (proc.time()[['elapsed']] > deadline) break
      Sys.sleep(0.005)
    }
    start = Sys.time()

    # Store error for re-throwing
    error_obj <- NULL

    tryCatch({
      result = invisible(eval(substitute(expr), parent.frame()))
    }, interrupt = function(i) {
      aborted <<- TRUE
      error_obj <<- i
    }, error = function(e) {
      aborted <<- TRUE
      error_obj <<- e
    }, finally = {
      duration = as.numeric(difftime(Sys.time(), start, units = 'secs'))
      curr_process$signal(15L)  # SIGTERM
      curr_process$wait(timeout = 5000)
      if (curr_process$is_alive()) curr_process$kill()
      stdout_output <- tryCatch(
        paste(readLines(tmpf, warn = FALSE), collapse = "\n"),
        error = function(e) "")
      unlink(tmpf)
      con <- textConnection(stdout_output)
      # Handle empty output
      df <- tryCatch({
        read.csv(con, header = FALSE, sep = ",", strip.white = TRUE,
          col.names = c("Integer", "Percentage"))
      }, error = function(e) {
        data.frame(Integer = 0, Percentage = 0.0)
      })
      close(con)
      # Ensure valid data
      if (nrow(df) == 0 || all(is.na(df$Integer))) {
        peak_mem <- 0
        percent <- 0.0
      } else {
        peak_mem <- max(df$Integer, na.rm = TRUE)
        percent <- max(df$Percentage, na.rm = TRUE)
        if (is.infinite(peak_mem)) peak_mem <- 0
        if (is.infinite(percent)) percent <- 0.0
      }

      new_memory <- round(peak_mem / 1024 / 1024, 2)
      new_percent <- round(percent, 2)

      if (message %in% names(env$timings)) {
        env$timings[[message]]$duration <-
          env$timings[[message]]$duration + duration
        env$timings[[message]]$max_mem <-
          max(env$timings[[message]]$max_mem, new_memory)
        env$timings[[message]]$mem_percent <-
          max(env$timings[[message]]$mem_percent, new_percent)
        env$timings[[message]]$aborted <-
          env$timings[[message]]$aborted || aborted
      } else {
        env$timings[[message]] = list(
          duration = duration,
          max_mem = new_memory,
          mem_percent = new_percent,
          aborted = aborted
        )
      }

      if (!silent) {
        time_str = format_time(duration)
        status = if (aborted) 'aborted after' else 'took'
        cat(sprintf(
          '%s %s %s using %.2f GiB\n\n',
          message, status, time_str, new_memory))
      }
      gc()
    })
    # Re-throw after cleanup, preserving the original condition class.
    # stop() only accepts error conditions; for interrupts use the
    # base calling handler approach.
    if (!is.null(error_obj)) {
      if (inherits(error_obj, "error")) {
        stop(error_obj)
      } else {
        # interrupt or other non-error condition
        signalCondition(error_obj)
      }
    }

    return(invisible(result))
  }

  format_time = function(duration, unit = NULL) {
    duration = as.numeric(duration)

    if (!is.null(unit)) {
      if (!(unit %in% names(.TIME_CONVERSIONS)))
        stop("Unsupported unit: ", unit)
      converted = duration * .TIME_CONVERSIONS[[unit]]
      return(paste0(format(converted, scientific = FALSE), unit))
    }

    units = list(
      list(threshold = 86400, suffix = 'd'),
      list(threshold = 3600, suffix = 'h'),
      list(threshold = 60, suffix = 'm'),
      list(threshold = 1, suffix = 's'),
      list(threshold = 0.001, suffix = 'ms'),
      list(threshold = 0.000001, suffix = 'µs'),
      list(threshold = 0.000000001, suffix = 'ns')
    )

    parts = c()
    for (unit in units) {
      threshold = unit$threshold
      suffix = unit$suffix
      if (duration >= threshold ||
          (length(parts) == 0 && threshold == 0.000000001)) {
        if (threshold >= 1) {
          value = as.integer(duration %/% threshold)
          duration = duration %% threshold
        } else {
          value = as.integer((duration / threshold) %% 1000)
        }
        if (value > 0 ||
            (length(parts) == 0 && threshold == 0.000000001)) {
          parts = c(parts, paste0(value, suffix))
        }
        if (length(parts) == 2) break
      }
    }

    if (length(parts) > 0) {
      paste(parts, collapse = ' ')
    } else {
      'less than 1ns'
    }
  }

  print_summary = function(sort = FALSE, unit = NULL) {
    cat('\n--- Timing Summary ---\n')

    if (length(env$timings) == 0) {
      cat('no timings recorded.\n')
      return(invisible(NULL))
    }

    if (sort) {
      durations = sapply(env$timings, function(x) x$duration)
      items = names(env$timings)[order(durations, decreasing = TRUE)]
    } else {
      items = names(env$timings)
    }

    total_time = sum(sapply(env$timings, function(x) x$duration))

    # Create table headers
    duration_header = if (!is.null(unit)) {
      paste0('Duration (', unit, ')')
    } else {
      'Duration'
    }
    headers = c('Operation', 'Status', duration_header, '% of Total',
                'Memory (GiB)', '% of Avail')

    # Create table data
    table_data = list()
    for (i in seq_along(items)) {
      msg = items[i]
      info = env$timings[[msg]]
      duration = info$duration
      percentage = if (total_time > 0) {
        (duration / total_time) * 100
      } else {
        0
      }
      max_mem = info$max_mem
      mem_percent = info$mem_percent
      status = if (info$aborted) 'aborted' else 'completed'
      time_str = format_time(duration, unit)

      table_data[[i]] = c(
        msg,
        status,
        time_str,
        sprintf('%.2f%%', percentage),
        sprintf('%.2f', max_mem),
        sprintf('%.2f%%', mem_percent)
      )
    }

    cat(sprintf('%-30s %-10s %-15s %-10s %-13s %-10s\n',
                headers[1], headers[2], headers[3], headers[4],
                headers[5], headers[6]))
    cat(paste(rep('-', 90), collapse = ''), '\n')

    for (row in table_data) {
      cat(sprintf('%-30s %-10s %-15s %-10s %-13s %-10s\n',
                  row[1], row[2], row[3], row[4], row[5], row[6]))
    }

    cat(sprintf('\nTotal time: %s\n', format_time(total_time, unit)))
  }

  to_dataframe = function(sort = TRUE, unit = NULL) {
    if (length(env$timings) == 0) {
      return(data.frame(
        operation = character(0), duration = numeric(0),
        duration_unit = character(0), aborted = logical(0),
        percentage = numeric(0), memory = numeric(0),
        memory_unit = character(0), percent_mem = numeric(0)
      ))
    }

    if (sort) {
      durations = sapply(env$timings, function(x) x$duration)
      items = names(env$timings)[order(durations, decreasing = TRUE)]
    } else {
      items = names(env$timings)
    }

    total = sum(sapply(env$timings, function(x) x$duration))

    ops = items
    durs = sapply(items, function(msg) env$timings[[msg]]$duration)

    if (!is.null(unit)) {
      if (!(unit %in% names(.TIME_CONVERSIONS)))
        stop("Unsupported unit: ", unit)
      durs = durs * .TIME_CONVERSIONS[[unit]]
      duration_unit = unit
    } else {
      duration_unit = "s"
    }

    aborts = sapply(items, function(msg) env$timings[[msg]]$aborted)
    pcts = sapply(items, function(msg) {
      if (total > 0) {
        (env$timings[[msg]]$duration / total) * 100
      } else {
        0
      }
    })
    memory = sapply(items, function(msg) env$timings[[msg]]$max_mem)
    memory_unit = rep("GiB", length(items))
    percent_mem = sapply(items, function(msg) {
      env$timings[[msg]]$mem_percent
    })

    data.frame(
      operation = ops,
      duration = durs,
      duration_unit = duration_unit,
      aborted = aborts,
      percentage = pcts,
      memory = memory,
      memory_unit = memory_unit,
      percent_mem = percent_mem
    )
  }

  shutdown = function() {
    if (env$.shutdown_done) return(invisible(NULL))
    env$.shutdown_done <- TRUE
    if (length(env$timings) == 0) return(invisible(NULL))
    print_summary(unit = env$.summary_unit)
    if (!is.null(env$.csv_path)) {
      df <- to_dataframe(sort = FALSE, unit = env$.unit)
      if (!is.null(env$.csv_columns)) {
        for (col_name in names(env$.csv_columns)) {
          df[[col_name]] <- env$.csv_columns[[col_name]]
        }
      }
      write.csv(df, env$.csv_path, row.names = FALSE)
    }
  }

  # Auto-flush timings on R exit (e.g. SLURM timeout)
  shutdown_ref <- new.env(parent = emptyenv())
  shutdown_ref$fn <- shutdown
  reg.finalizer(shutdown_ref, function(e) e$fn(), onexit = TRUE)

  structure(list(
    with_timer = with_timer,
    print_summary = print_summary,
    to_dataframe = to_dataframe,
    shutdown = shutdown
  ), class = "TimerCollection")
}

system_info <- function() {
  hostname <- Sys.info()['nodename']
  user <- Sys.getenv("USER")
  if (user == "") user <- "N/A"

  # Get CPU cores
  cpu_task <- Sys.getenv("SLURM_CPUS_PER_TASK")
  cpu_cores <- Sys.getenv("SLURM_CPUS_ON_NODE")

  if (nchar(cpu_task) > 0) {
    cpu_cores <- cpu_task
  } else if (nchar(cpu_cores) == 0) {
    tryCatch({
      if (requireNamespace("parallel", quietly = TRUE)) {
        cpu_cores <- parallel::detectCores()
      } else {
        cpu_cores <- "N/A"
      }
    }, error = function(e) {
      cpu_cores <<- "N/A"
    })
  }

  # Get Memory
  mem_gb_str <- "N/A"
  mem_mb_str <- Sys.getenv("SLURM_MEM_PER_NODE")

  if (nchar(mem_mb_str) > 0) {
    # First, try SLURM environment variable
    mem_mb <- suppressWarnings(as.numeric(mem_mb_str))
    if (!is.na(mem_mb)) {
      mem_gb_str <- sprintf("%.1f GB", mem_mb / 1024)
    }
  } else {
    # As a fallback, read from /proc/meminfo
    tryCatch({
      meminfo_lines <- readLines("/proc/meminfo")
      for (line in meminfo_lines) {
        if (grepl("^MemTotal:", line)) {
          mem_kb <- as.numeric(strsplit(line, "\\s+")[[1]][2])
          mem_gb_str <- sprintf("%.1f GB", mem_kb / 1024 / 1024)
          break
        }
      }
    }, error = function(e) {
      # Keep default "N/A"
    })
  }

  cat("\n--- User Resource Allocation ---\n")
  cat(sprintf("Node: %s\n", hostname))
  cat(sprintf("User: %s\n", user))
  cat(sprintf("CPU Cores Allocated: %s\n", cpu_cores))
  cat(sprintf("Memory Allocated: %s\n", mem_gb_str))
  cat(sprintf("R.version=%s\n", R.version.string))
  cat("\n")
}

read_h5ad_obs <- function(path) {
  # Load hdf5r only when needed
  if (!requireNamespace("hdf5r", quietly = TRUE)) {
    stop("Package 'hdf5r' required for reading h5ad files")
  }

  h5_file <- hdf5r::H5File$new(path, mode = "r")
  on.exit(h5_file$close_all(), add = TRUE)

  obs_group <- h5_file[["obs"]]

  index_attr <- obs_group$attr_open("_index")
  index_col_name <- index_attr$read()
  index_attr$close()

  index <- obs_group[[index_col_name]][]

  all_obs_items <- obs_group$names
  data_cols <- setdiff(all_obs_items, c("__categories", index_col_name))

  obs_list <- list()

  if ("__categories" %in% all_obs_items) {
    categories_group <- obs_group[["__categories"]]
    category_names <- categories_group$names

    for (col in data_cols) {
      item <- obs_group[[col]]
      if (inherits(item, "H5D") && item$dims == length(index)) {
        if (col %in% category_names) {
          codes <- item[]
          levels <- categories_group[[col]][]
          codes_r <- codes + 1L
          codes_r[codes_r <= 0L] <- NA_integer_
          obs_list[[col]] <- factor(levels[codes_r], levels = levels)
        } else {
          obs_list[[col]] <- item[]
        }
      }
    }
  } else {
    for (col in data_cols) {
      item <- obs_group[[col]]
      if (inherits(item, "H5Group") &&
          all(c("codes", "categories") %in% item$names)) {
          codes <- item[["codes"]][]
          levels <- item[["categories"]][]
          codes_r <- codes + 1L
          codes_r[codes_r <= 0L] <- NA_integer_
          obs_list[[col]] <- factor(levels[codes_r], levels = levels)
      } else if (inherits(item, "H5D") && item$dims == length(index)) {
          obs_list[[col]] <- item[]
      }
    }
  }

  obs_df <- as.data.frame(obs_list, check.names = FALSE)
  rownames(obs_df) <- index

  return(obs_df)
}

transfer_accuracy <- function(obs_df, orig_col, trans_col) {
  accuracy_df <- obs_df %>%
    mutate(
      orig = as.character(.data[[orig_col]]),
      trans = as.character(.data[[trans_col]])) %>%
    group_by(orig) %>%
    summarize(
      n_correct = sum(orig == trans),
      n_total = n(),
      .groups = "drop") %>%
    bind_rows(
      summarize(
        .,
        orig = "Total",
        n_correct = sum(n_correct),
        n_total = sum(n_total))) %>%
    mutate(percent_correct = (n_correct / n_total) * 100) %>%
    rename(cell_type = orig)
  print(accuracy_df, n = Inf)
  return(accuracy_df)
}

File: utils_local.py
import gc
import io
import os
import sys
import time
import socket
import tempfile
import subprocess
import numpy as np
import polars as pl
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from timeit import default_timer
from contextlib import contextmanager
from typing import ContextManager
from tabulate import tabulate

def plural(string: str, count: int) -> str:
    return string if abs(count) == 1 else f'{string}s'

def run(cmd, *, log_file=None, unbuffered=False, pipefail=True,
        num_threads=None, **kwargs):
    run_kwargs = dict(check=True, shell=True, executable='/bin/bash')
    run_kwargs.update(**kwargs)
    thread_str = (
        f'export MKL_NUM_THREADS={num_threads}; '
        f'export OMP_NUM_THREADS={num_threads}; '
        f'export OPENBLAS_NUM_THREADS={num_threads}; '
        f'export NUMEXPR_MAX_THREADS={num_threads}; '
        if num_threads is not None else '')
    return subprocess.run(
        f'{thread_str}'
        f'set -eu{"o pipefail" if pipefail else ""}; '
        f'{"stdbuf -i0 -o0 -e0 " if unbuffered else ""}{cmd}'
        f'{f" 2>&1 | tee {log_file}" if log_file is not None else ""}',
        **run_kwargs)

_MONITOR_MEM_SH_PATH = os.path.join(
    os.path.dirname(__file__), 'monitor_mem.sh')

POLLING_INTERVAL = '0.05'

class MemoryTimer:
    _TIME_CONVERSIONS = {
        's': 1, 'ms': 1000, 'us': 1e6, 'µs': 1e6,
        'ns': 1e9, 'm': 1/60, 'h': 1/3600, 'd': 1/86400
    }

    def __init__(self, silent=True, csv_path=None, csv_columns=None,
                 unit='s', summary_unit=None):
        import atexit
        import signal

        self.timings = {}
        self.silent = silent
        self._csv_path = csv_path
        self._csv_columns = csv_columns or {}
        self._unit = unit
        self._summary_unit = summary_unit or unit
        self._shutdown_done = False

        def _sigterm_handler(signum, frame):
            signal.signal(signal.SIGTERM, signal.SIG_IGN)
            signal.signal(signal.SIGABRT, signal.SIG_IGN)
            raise SystemExit(1)

        signal.signal(signal.SIGTERM, _sigterm_handler)
        signal.signal(signal.SIGABRT, _sigterm_handler)
        atexit.register(self.shutdown)

    def __call__(self, message: str, exclude=False) -> ContextManager[None]:
        pid = os.getpid()
        @contextmanager
        def timer():
            if not self.silent:
                print(f'{message}...', flush=True)
            monitor, tmpf_path = self._start_monitor(pid)
            start = default_timer()
            aborted = False
            try:
                yield
            except BaseException:
                aborted = True
                raise
            finally:
                duration = default_timer() - start
                memory_gb, percent_mem = self._stop_monitor(monitor, tmpf_path)
                if not self.silent:
                    status = 'aborted after' if aborted else 'took'
                    time_str = self._format_time(duration)
                    print(f'{message} {status} {time_str} using '
                          f'{memory_gb} GiB\n', flush=True)
                self._record(message, duration, memory_gb, percent_mem,
                             aborted, exclude)
                gc.collect()

        return timer()

    def _start_monitor(self, pid):
        # Write monitor output to a temp file instead of a pipe.
        # The 64 KB pipe buffer fills during long operations (the
        # main thread is busy and never reads), blocking the monitor
        # and missing late peaks.
        tmpf = tempfile.NamedTemporaryFile(
            prefix='memmon_', suffix='.csv', delete=False)
        tmpf.close()
        stdout_file = open(tmpf.name, 'w')
        monitor = subprocess.Popen(
            [_MONITOR_MEM_SH_PATH, '-p', str(pid), '-i', POLLING_INTERVAL],
            stdout=stdout_file
        )
        stdout_file.close()
        # Sync: wait for first sample (file must have content)
        deadline = time.monotonic() + 5
        while time.monotonic() < deadline:
            if os.path.getsize(tmpf.name) > 0:
                break
            time.sleep(0.005)
        return monitor, tmpf.name

    def _stop_monitor(self, monitor, tmpf_path):
        try:
            monitor.terminate()
            monitor.wait(timeout=5)
        except ProcessLookupError:
            pass
        except subprocess.TimeoutExpired:
            monitor.kill()
            monitor.wait()
        try:
            with open(tmpf_path) as f:
                output = f.read().strip()
        except OSError:
            output = ''
        finally:
            try:
                os.unlink(tmpf_path)
            except OSError:
                pass

        if not output:
            data = np.array([[0.0, 0.0]])
        else:
            try:
                data = np.loadtxt(io.StringIO(output), delimiter=',')
            except ValueError:
                data = np.array([[0.0, 0.0]])

        if data.size == 0:
            data = np.array([[0.0, 0.0]])
        elif data.ndim == 1:
            data = data[None, :]

        max_mem = np.max(data, axis=0)
        memory_gb = np.round(max_mem[0] / 1024 / 1024, 2)
        percent_mem = np.round(max_mem[1], 2)
        return memory_gb, percent_mem

    def _record(self, message, duration, memory_gb, percent_mem, aborted, exclude=False):
        if message in self.timings:
            self.timings[message]['duration'] += duration
            self.timings[message]['memory'] = max(
                self.timings[message]['memory'], memory_gb)
            self.timings[message]['%mem'] = max(
                self.timings[message]['%mem'], percent_mem)
            self.timings[message]['aborted'] = (
                self.timings[message]['aborted'] or aborted)
        else:
            self.timings[message] = {
                'duration': duration,
                'memory': memory_gb,
                '%mem': percent_mem,
                'aborted': aborted,
                'exclude': exclude,
            }

    def shutdown(self):
        # Write timing summary and CSV. No-op if already called.
        if self._shutdown_done:
            return
        self._shutdown_done = True
        if not self.timings:
            return
        self.print_summary(unit=self._summary_unit)
        if self._csv_path:
            df = self.to_dataframe(sort=False, unit=self._unit)
            if self._csv_columns:
                df = df.with_columns(
                    [pl.lit(v).alias(k)
                     for k, v in self._csv_columns.items()])
            df.write_csv(self._csv_path)

    def print_summary(self, sort=False, unit=None):
        print('\n--- Timing Summary ---')
        items = (sorted(self.timings.items(),
                        key=lambda x: x[1]['duration'], reverse=True)
                 if sort else list(self.timings.items()))

        total_time = sum(info['duration'] for _, info in items
                         if not info.get('exclude'))
        table_data = []
        duration_header = f'Duration ({unit})' if unit else 'Duration'
        headers = [
            'Operation', 'Status', duration_header,
            '% of Total', 'Memory (GiB)', '% of Avail']

        for message, info in items:
            duration = info['duration']
            memory = info['memory']
            pct = (duration / total_time * 100) if total_time else 0
            status = 'aborted' if info['aborted'] else 'completed'
            time_str = self._format_time(duration, unit)
            table_data.append([
                message,
                status,
                time_str,
                'N/A' if info.get('exclude') else f'{pct:.2f}%',
                f'{memory}',
                'N/A' if info.get('exclude') else f'{info["%mem"]:.2f}%'
            ])

        print(tabulate(table_data, headers=headers, tablefmt='simple'))
        print(f'\nTotal time: {self._format_time(total_time, unit)}')

    def _format_time(self, duration, unit=None):
        if unit:
            if unit not in self._TIME_CONVERSIONS:
                raise ValueError(f'Unsupported unit: {unit}')
            return f'{duration * self._TIME_CONVERSIONS[unit]:.2f}{unit}'

        units = [
            (86400, 'd'), (3600, 'h'), (60, 'm'), (1, 's'),
            (0.001, 'ms'), (1e-6, 'µs'), (1e-9, 'ns')
        ]

        parts = []
        for threshold, suffix in units:
            if duration >= threshold or (not parts and suffix == 'ns'):
                if threshold >= 1:
                    value = int(duration // threshold)
                    duration %= threshold
                else:
                    value = int((duration / threshold) % 1000)
                if value > 0 or (not parts and suffix == 'ns'):
                    parts.append(f'{value}{suffix}')
                if len(parts) == 2:
                    break

        return ' '.join(parts) if parts else 'less than 1ns'

    def to_dataframe(self, sort=True, unit=None):
        if not self.timings:
            return pl.DataFrame({
                'operation': [], 'duration': [], 'duration_unit': [],
                'aborted': [], 'percentage': [], 'memory': [],
                'memory_unit': [], 'percent_mem': []
            })

        items = (sorted(self.timings.items(),
                        key=lambda x: x[1]['duration'], reverse=True)
                 if sort else list(self.timings.items()))

        total = sum(info['duration'] for info in self.timings.values()
                    if not info.get('exclude'))

        conv = 1.0
        if unit:
            if unit not in self._TIME_CONVERSIONS:
                raise ValueError(f'Unsupported unit: {unit}')
            conv = self._TIME_CONVERSIONS[unit]

        rows = [
            {
                'operation': msg,
                'duration': info['duration'] * conv if unit else info['duration'],
                'duration_unit': unit or 's',
                'aborted': info['aborted'],
                'percentage': None if info.get('exclude') else ((info['duration'] / total * 100) if total else 0),
                'memory': info['memory'],
                'memory_unit': 'GiB',
                'percent_mem': None if info.get('exclude') else info['%mem'],
            }
            for msg, info in items
        ]

        return pl.DataFrame(rows)

def print_df(df, num_rows=-1, num_columns=-1):
    with pl.Config(tbl_rows=num_rows, tbl_cols=num_columns):
        print(df)

def system_info():
    hostname = socket.gethostname()
    user = os.environ.get('USER', 'N/A')
    cpu_cores = (os.environ.get('SLURM_CPUS_PER_TASK') or
                 os.environ.get('SLURM_CPUS_ON_NODE'))
    if not cpu_cores:
        try:
            cpu_cores = os.cpu_count()
        except NotImplementedError:
            cpu_cores = 'N/A'

    mem_gb = 'N/A'
    try:
        mem_mb = int(os.environ['SLURM_MEM_PER_NODE'])
        mem_gb = f'{mem_mb / 1024:.1f} GB'
    except (KeyError, ValueError):
        try:
            with open('/proc/meminfo', 'r') as f:
                for line in f:
                    if line.startswith('MemTotal:'):
                        mem_kb = int(line.split()[1])
                        mem_gb = f'{mem_kb / 1024 / 1024:.1f} GB'
                        break
        except FileNotFoundError:
            pass

    print(f'\n--- User Resource Allocation ---', flush=True)
    print(f'Node: {hostname}', flush=True)
    print(f'User: {user}', flush=True)
    print(f'CPU Cores Allocated: {cpu_cores}', flush=True)
    print(f'Memory Allocated: {mem_gb}', flush=True)
    print(f'Python Version: {sys.version}', flush=True)

    gpu_info = 'N/A'
    try:
        result = subprocess.run(
            ['nvidia-smi', '--query-gpu=name,memory.total',
             '--format=csv,noheader'],
            capture_output=True, text=True, timeout=5)
        if result.returncode == 0 and result.stdout.strip():
            gpu_info = result.stdout.strip()
    except (FileNotFoundError, subprocess.TimeoutExpired):
        pass
    print(f'GPU: {gpu_info}', flush=True)

    print(flush=True)

def run_slurm(cmd, *, job_name='job', log_file=None,
              CPUs=192 if os.environ.get('CLUSTER') == 'trillium' else
                   96 if os.environ.get('CLUSTER') == 'trillium-gpu' else 1,
              GPUs=4 if os.environ.get('CLUSTER') == 'trillium-gpu' else 0,
              days=None, hours=None, memory=None,
              partition='compute' if os.environ.get('CLUSTER') == 'trillium'
                         else 'compute_full_node' if os.environ.get('CLUSTER') == 'trillium-gpu'
                         else None,
              account=None, verbose=False):
    cluster = os.environ.get('CLUSTER')
    runtime = f'{days}-00:00:00' if days is not None else \
        f'{hours}:00:00' if hours is not None else '1-00:00:00' \
            if partition != 'debug' else '1:00:00'
    if verbose:
        memory_description = f' and {memory[:-1]} {memory[-1]}iB memory' \
            if memory is not None else ''
        partition_description = \
            f' on the {partition} partition' if partition is not None else ''
        print(f'Requesting {CPUs} {plural("CPU", CPUs)}, {GPUs} '
              f'{plural("GPU", GPUs)}{memory_description} for {runtime}'
              f'{partition_description}...')
    job_name = job_name.replace(' ', '_')
    from tempfile import NamedTemporaryFile
    try:
        with NamedTemporaryFile('w', dir=os.environ.get('SCRATCH', '.'),
                                suffix='.sh', delete=False) as temp_file:
            partition_settings = f'#SBATCH -p {partition}\n' \
                if partition is not None else ''
            account_settings = f'#SBATCH --account={account}\n' \
                if account is not None else ''
            cpu_settings = '' if cluster.startswith('trillium') else \
                f'#SBATCH -c {CPUs}\n'
            gpu_settings = f'#SBATCH --gpus-per-node=h100:{GPUs}\n' \
                if GPUs > 0 else ''
            memory_settings = f'#SBATCH --mem {memory}\n' \
                if memory is not None else ''
            print(
                f'#!/bin/bash\n'
                f'{partition_settings}'
                f'{account_settings}'
                f'#SBATCH -N 1\n'
                f'#SBATCH -n 1\n'
                f'{gpu_settings}'
                f'{cpu_settings}'
                f'{memory_settings}'
                f'#SBATCH -t {runtime}\n'
                f'#SBATCH -J {job_name}\n'
                f'{f"#SBATCH -o {log_file}" if log_file is not None else ""}\n'
                f'#SBATCH --signal=B:TERM@30\n'
                f'export PYTHONUNBUFFERED=1\n'
                f'export R_LIBS_USER=/home/wainberg/R/x86_64-pc-linux-gnu-library/4.4\n'
                f'export OMP_PLACES=cores\n'
                f'export OMP_PROC_BIND=spread\n'
                f'export HDF5_USE_FILE_LOCKING=FALSE\n'
                f'set -m\n'
                f'{cmd} &\n'
                f'CHILD=$!\n'
                f'trap "kill -INT -$CHILD 2>/dev/null; '
                f'wait $CHILD 2>/dev/null; exit \\$?" TERM\n'
                f'wait $CHILD\n'
                f'exit $?\n',
                file=temp_file)
        sbatch = '.sbatch' if cluster.startswith('trillium') else 'sbatch'
        sbatch_message = run(f'{sbatch} {temp_file.name}',
                             stdout=subprocess.PIPE)\
            .stdout.decode().rstrip('\n')
        print(f'{sbatch_message} ("{job_name}")')
    finally:
        try:
            os.unlink(temp_file.name)
        except NameError:
            pass

def transfer_accuracy(obs, orig_col, trans_col):
    if isinstance(obs, pd.DataFrame):
        obs = pl.from_pandas(obs)
    groups = obs\
        .with_columns(pl.col(orig_col, trans_col).cast(pl.String))\
        .group_by(orig_col)\
        .agg(n_correct=pl.col(orig_col).eq(pl.col(trans_col)).sum(),
             n_total=pl.len())
    df = pl.concat([
            groups,
            groups.sum().with_columns(pl.lit('Total').alias(orig_col))])\
        .with_columns(
            percent_correct=pl.col('n_correct') / pl.col('n_total') * 100)\
        .sort(pl.col(orig_col).eq('Total'), orig_col)\
        .rename({orig_col: 'cell_type'})
    print_df(df)
    return df

def confusion_matrix_plot(sc_obs, orig_col, trans_col, filename):
    cm = pd.crosstab(
        sc_obs[orig_col].to_pandas(),
        sc_obs[trans_col].to_pandas()
    )
    with pd.option_context('display.max_columns', None):
        print(cm)

    norm_cm = cm.div(cm.sum(axis=1), axis=0).fillna(0)

    height = 6 + norm_cm.shape[0] * 0.05
    width = 8 + norm_cm.shape[1] * 0.05

    plt.figure(figsize=(width, height))
    ax = sns.heatmap(
        norm_cm, square=False, linewidths=0.5,
        cmap='rocket_r', vmin=0, vmax=1,
        cbar_kws={'shrink': 0.7}
    )
    cbar = ax.collections[0].colorbar
    cbar.set_ticks([0, 0.2, 0.4, 0.6, 0.8, 1])
    cbar.set_ticklabels(['0%', '20%', '40%', '60%', '80%', '100%'])

    plt.xticks(rotation=90)
    plt.yticks(rotation=0)
    ax.set_xlabel(f'Predicted: {trans_col}')
    ax.set_ylabel(f'Original: {orig_col}')
    plt.tight_layout()
    plt.savefig(filename, dpi=300)
    plt.show()
